projects/chaos/src/sequencer.c

```diff
@@ -0,0 +1,231 @@
+#include "sequencer.h"
+
+#include <stddef.h>
+#include <string.h>
+
+#include "chaos_events.h"
+#include "event_queue.h"
+#include "fsm.h"
+#include "log.h"
+#include "misc.h"
+#include "power_path.h"
+#include "relay_id.h"
+
+#define EMPTY_DATA 0
+
+static FSM s_sequencer_fsm;
+static const Event *s_events;
+static uint8_t s_retries = 0;
+static uint16_t s_event_idx;
+static uint16_t s_sizeof_events = 0;
+
+FSM_DECLARE_STATE(sequencer_state_emergency);
+FSM_DECLARE_STATE(sequencer_state_idle);
+FSM_DECLARE_STATE(sequencer_state_charge);
+FSM_DECLARE_STATE(sequencer_state_drive);
+
+FSM_STATE_TRANSITION(sequencer_state_emergency) {
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_IDLE, sequencer_state_idle);
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_EMERGENCY, sequencer_state_emergency);
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_RESET, sequencer_state_emergency);
+}
+
+FSM_STATE_TRANSITION(sequencer_state_idle) {
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_EMERGENCY, sequencer_state_emergency);
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_CHARGE, sequencer_state_charge);
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_DRIVE, sequencer_state_drive);
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_RESET, sequencer_state_idle);
+}
+
+FSM_STATE_TRANSITION(sequencer_state_charge) {
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_EMERGENCY, sequencer_state_emergency);
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_IDLE, sequencer_state_idle);
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_RESET, sequencer_state_charge);
+}
+
+FSM_STATE_TRANSITION(sequencer_state_drive) {
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_EMERGENCY, sequencer_state_emergency);
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_IDLE, sequencer_state_idle);
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_RESET, sequencer_state_drive);
+}
+
+static void prv_sequencer_state_emergency(FSM *fsm, const Event *e, void *context) {
+  (void)fsm;
+  (void)e;
+  (void)context;
```

Comment:
Might as well use the context pointers since they exist already. I know you only plan on having one sequencer so I can understand why it exists in this module, but I'm not a huge fan of interacting directly with static variables.

Reply:
Done.